### PR TITLE
chore(docs): create script for generating Cue documentation output from configuration schema

### LINKED ISSUE #14814: chore(config, docs): write a script to generate equivalent Cue output from the configuration schema
As part of the work for #14812, we need to first write a script that takes the resulting configuration schema for a version of Vector, and converts that into (near-)equivalent Cue output, suitable for building the HTML documentation with.

## Main Outcomes

- A simple script exists within the repository for doing this, that all developers can run. (works both locally and in CI, basically)
- It can do all steps in a single run: read the schema, find all the component definitions, etc, parse the schema and convert it into Cue of an equivalent shape to what exists now
- It writes this Cue output to the right location, but doesn't overwrite existing files. It should live side-by-side for now.

### PR DESCRIPTION
**Pull request by tobz**
This PR introduces a new script, `generate-components-docs.rb`, which parses Vector configuration schema output and generates comparable Cue output to match our existing Cue-based documentation.

This script currently places the generated Cue output in "base" folders in preparation for work in #14815, but we do not yet include those files in the repository as they still need output tweaks to get their shape closer to the existing Cue documentation.

There's also an accompanying Makefile target -- `generate-component-docs` -- which builds Vector and generates the configuration schema file, used to feed `generate-components-docs.rb`, before calling the script.

## Reviewer notes

I've had to make a bunch of changes on the Rust side to get the configuration schema into the right shape, or to curry enough information for `generate-components-docs.rbi` to do its thing. This is a non-exhaustive list, in no particular order, of those changes:

- tweaked the title/description of buffer-related types as the normal code doc comments were too verbose/didn't match our intended external-facing docs
- added new metadata to all generated enum schemas (as well as manually backfilled this for manual `Configurable` impls) to describe the tagging mode being used by `serde`, which allows us to better reconstruct the human-friendly description of fields using these types
- fixed a glitch in the schema output when all fields on a struct were marked `#[serde(flatten)]`, which lead to an extraneous empty object subschema in the `allOf` condition
- added metadata to numeric types to signify their subtype i.e. instead of just "number", we can now differentiate signed vs unsigned vs floating point
- added metadata to the component "outer" types to let `generate-components-docs.rb` understand which types are the common "base" for a given component type vs what is an actual individual instance of that component type

Closes #14814.

### CODE REVIEW

File: scripts/generate-components-docs.rb [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+        # It's hard to propagate the default from the configuration schema generation side, but much
+        # easier to do so here.
+        single_subschema = deep_copy(subschemas[single_idx])
+        if get_schema_type(single_subschema) == json_type_str(schema['default'])
+          single_subschema['default'] = schema['default']
+        end
+        resolved_subschema = resolve_schema(root_schema, subschemas[single_idx])
+
+        @logger.debug "Resolved as 'X or array of X' enum schema."
+        # @logger.debug "Original schema: #{schema}"
```
Comment (spencergilbert) [inline scripts/generate-components-docs.rb:488]:
```suggestion
```

Removed commented line?

File: scripts/generate-components-docs.rb [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+          'string' => {
+            'enum' => unique_tag_values.transform_values do |tag_schema|
+                        get_rendered_description_from_schema(tag_schema)
+                      end
+          }
+        }
+      }
+
+      @logger.debug "Resolved as 'internally-tagged with named fields' enum schema."
+      # @logger.debug "Original schema: #{schema}"
```
Comment (spencergilbert) [inline scripts/generate-components-docs.rb:629]:
```suggestion
```

Remove commented line?

File: lib/vector-buffers/src/config.rs [hunk 3]
```diff
@@ -194,7 +194,11 @@ impl DiskUsage {
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 #[serde(rename_all = "snake_case", tag = "type")]
 pub enum BufferType {
-    /// A buffer stage backed by an in-memory channel provided by `tokio`.
+    /// A buffer stage backed by an in-memory channel provided by `tokio``g.
```
[the remark this answers is not in the document]
Comment (tobz) [reply lib/vector-buffers/src/config.rs:197]: Good catch. Will fix.

File: scripts/generate-components-docs.rb (hunk 1 above)
Comment (tobz) [reply scripts/generate-components-docs.rb:488]: Yeah, good call. Will remove.

File: scripts/generate-components-docs.rb (hunk 2 above)
Comment (tobz) [reply scripts/generate-components-docs.rb:629]: Yeah, good call. Will remove.

**Review by spencergilbert — APPROVED** [review]
I can't necessarily speak to the contents/writing of the ruby script. Rust looks looks fine, and if the output of the script is 👍 - good enough for me

File: lib/vector-buffers/src/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tobz) [reply lib/vector-buffers/src/config.rs]: Addressed in 839b4ef30.

File: scripts/generate-components-docs.rb (hunk 2 above)
Comment (tobz) [reply scripts/generate-components-docs.rb]: Addressed in 839b4ef30.

File: scripts/generate-components-docs.rb (hunk 1 above)
Comment (tobz) [reply scripts/generate-components-docs.rb]: Addressed in 839b4ef30.

**Pull request merged**